Kibana query language autocomplete

lukasolson opened the pull request:
Replaces #14523.
Closes #12315.
Relies on https://github.com/elastic/kibana/pull/17069.

This PR adds autocomplete functionality to the [simplified Kibana query language](https://github.com/elastic/kibana/pull/15857). It will autocomplete field names, operators, values, and conjunctions.

Value suggestions can be turned off using the same flag to turn off filter editor value suggestions (`filterEditor:suggestValues`).

![mar-21-2018 16-26-35](https://user-images.githubusercontent.com/1178348/37742882-c162ddd6-2d24-11e8-8adb-e3ef7a38d2cc.gif)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15273/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15458/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15472/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15501/)

AlonaNadler commented:
Nice progress @lukasolson (i played with Daves branch), the design is really nice!

- The first autocomplete suggestion should be  highlighted/focused (default selection) - otherwise, the interaction requires to use the arrows too much
- We should limit the number of suggestions to maybe 10, right now there are too many…
- It only suggests fields that start with the string I’m typing. E.g “RelatedContent.twitter:card”  is not part of the suggestions when I type “twitter”

- Does it suggest a full condition (field:value)? relatedContent.og:url ? if so why does it suggest values if there are no results matching relatedContent.og:url
- When I choose .raw field follow by the operator it doesn’t suggest the possible values, it still shows the operators, only after I type something it starts suggesting values. I think immediately after the user selects an operator the values suggestions should appear (it might just take too long for the suggestions to appear).
- Spaces between fields, values, operators - how does it work?
- Suggestions start being relevant to the second character, I think it should at least show suggestion that start with the first character otherwise it’s confusing since I don't understand why it suggests these values
- We might have a problem with the speed of the suggestions, especially for values but also when you delete a character the suggestions take a long time to update
- Having the value suggestions only on the .raw fields makes values autocomplete less useful - I don't remember why we can’t suggest values for keyword fields by taking the suggestions from the .raw fields without changing how the query runs
- and/or should are not suggested 
- There is no way to know that you can search for list of values for a specific field without rewriting the field name, maybe we should re-evaluate this syntax

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/11/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/12/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/91/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/126/)

AlonaNadler commented:
@lukasolson  just pulled, there is an overlap in the lucene/kuery selector 
![image](https://user-images.githubusercontent.com/32550316/36566402-404fbb8a-17d8-11e8-8775-61bf8ba65519.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/172/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/192/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/204/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/226/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/232/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/271/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/273/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/277/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/279/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/289/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/291/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/295/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/374/)

lukasolson commented:
@AlonaNadler:

> When I type ip it only suggests clientip and not the ip field

I had put this in by design, seemed strange to suggest what's already typed, but I can also see why it'd be confusing if you didn't. I've gone back to the previous behavior where it still shows the fields even if they're an exact match.

> Sometimes it suggests only .raw fields

This was also related to the above, should be fixed now.

> What happened to the icons on autocomplete suggestions? it only have colors now

Since I'm relying on a new version of EUI, when you check out this PR you'll have to make sure and run `yarn kbn bootstrap` before starting up Kibana. Then you should see the icons.

> I think it looks a bit messy with the left side suggestion cut in the middle and showing the full string in the middle

@snide, what are your thoughts here? Should we extend the value out more, or perhaps use ellipsis?

> For exists operator, any chance we can this part more visible its too easy to miss

I'm not exactly sure what you mean... Is that operator somehow less visible than the other operators? I could move it up next to the `:` operator if that helps.

> Do we need the double quotes in every suggestion it just looks too busy

We should probably zoom about this one. Currently, the quotes signifies a phrase query (i.e. "I want it to match exactly"). If we don't include the quotes, it will still work in most cases but also might cause strange behavior (i.e. matching on documents that don't contain the exact phrase). It comes down to what we want the behavior to be.

> It doesn't always suggest and and or, is it intentionally?

I'm trying to improve this behavior and have made significant improvements since a couple of days ago. Could you try again and see if you still see this?

Thanks for the thorough feedback!

w33ble commented on src/ui/public/kuery/suggestions/index.js:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    const { suggestionTypes = [] } = cursorNode;
+    return Promise.all([...suggestionTypes, 'recentSearch'].map(type => {
```
Putting `recentSearch` first seems more natural. Otherwise, when you focus the query bar, you just get a huge list of fields and need to scroll down (with the mouse) to see recent searches:

![screenshot 2018-03-09 14 01 55](https://user-images.githubusercontent.com/404731/37229879-c2f4de14-23a2-11e8-996e-2cc95d99f3b0.png)

Versus:

![screenshot 2018-03-09 14 08 40](https://user-images.githubusercontent.com/404731/37230094-7bb3f980-23a3-11e8-89f7-039bdc1dd7d1.png)

lukasolson commented on src/ui/public/kuery/suggestions/conjunction.js:
```diff
@@ -0,0 +1,22 @@
+
+const conjunctions = {
+  and: `<p>Requires that <span class="suggestionItem__callout">both arguments</span> joined together both return true</p>`,
+  or: `<p>Requires <span class="suggestionItem__callout">only one of</span> the arguments joined together must return true</p>`
```
Good call, I'll update the verbiage there.

w33ble commented:
I find it odd that clicking on a previous query doesn't do anything more than update the input:

![mar-09-2018 14-16-56](https://user-images.githubusercontent.com/404731/37230529-12ab297a-23a5-11e8-8771-63a46bcae164.gif)

Likewise, selecting it with the keyboard and hitting enter also just updates the input:

![mar-09-2018 14-19-03](https://user-images.githubusercontent.com/404731/37230549-216dc198-23a5-11e8-97af-ef3e2fcca060.gif)

I expected that in both cases, it should close the the auto-complete dialog and execute the query. The requirement to click the search button or hit enter again seems odd. I could understand if I hit tab that it would just fill in the value without submitting the query, but enter really seems like it should submit.

---

Also related, it's odd to me that when the form *is* submitted that the auto-complete dialog doesn't go away.

![mar-09-2018 14-27-06](https://user-images.githubusercontent.com/404731/37230846-0cc4d73a-23a6-11e8-8e1d-20634ad7d3cb.gif)

w33ble commented:
Keyboard navigation will need to cause the option container to scroll up and down. Right now if you move beyond the last visible item in the list, the selection is simply hidden (latest Chrome):

![mar-09-2018 14-30-50](https://user-images.githubusercontent.com/404731/37230988-a56877ee-23a6-11e8-9b9d-b79867989a86.gif)

w33ble commented:
The hide/show state seems to be preserved on the input. Initially it's hidden, until the user types something, which seems correct. When the input loses focus, however, the show state is restored immediately when it gains focus again.

![mar-09-2018 14-43-30](https://user-images.githubusercontent.com/404731/37231401-470d42ae-23a8-11e8-92a0-b501c8cfd85f.gif)

As a result, mixed with the fact that it doesn't close when I press enter to submit the query, I find myself needing to hit escape a lot to close it manually.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/477/)

snide commented:
The up/down keyboard stuff can be fixed by making this stuff use actual `focus` for selection so that the browser will adjust against the selection, which has the secondary benefit of likely fixing a lot of the accessibility problems that are set here.

As to...

> I think it looks a bit messy with the left side suggestion cut in the middle and showing the full string in the middle

Can you give me a screen of your issue @AlonaNadler. It's likely something I can address.

lukasolson commented:
> The up/down keyboard stuff can be fixed by making this stuff use actual `focus` for selection so that the browser will adjust against the selection, which has the secondary benefit of likely fixing a lot of the accessibility problems that are set here.

If we actually use the focus, wouldn't that make it so that when the user starts typing, the query bar won't receive that input? I'm not really sure how to get around that.

snide commented:
@lukasolson That's a totally true point that I'm laughing I missed.

AlonaNadler commented:
@dave 
![image](https://user-images.githubusercontent.com/32550316/37234472-e23bfdfc-23ac-11e8-8577-bfd25e9bbef7.png)
- Can we make equal and exists bolder, it's too verbose right now and I think having the syntax definition bolder will be a better experience `:*` doesn't say anything to me at the inception `Exists` helps me more until users pick up the syntax

@lukas looks awesome, few last items (I promise) 
- joes comment regarding search history before suggestions, I would say it depends on the use case, you don't necessarily repeat your queries - I suggest let's take Google's approach - suggestions before history
- can  limit the number of suggestions, 
![image](https://user-images.githubusercontent.com/32550316/37234655-f3e9b99e-23ad-11e8-8ede-7db5846cf744.png)
- Any chance the suggestions won't be case sensitive e.g. when typing `Win 7` instead of `win 7` might cause confusion `geo.dest : CN` vs `geo.dest : cn` - I kinda know what you going to tell me ... :)
- Maybe worth considering hovering on the first suggestion as default, currently to select the suggestions I need to use the arrows and return, if I want to select the first suggestion intuitively I click return (to select it ) but then it triggers the search

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/526/)

Bargs commented on src/ui/public/query_bar/directive/suggestion.html:
```diff
@@ -0,0 +1,23 @@
+    </div>
+  </div>
+  <div class="suggestionItem__text" ng-bind-html="item.text"></div>
+  <div class="suggestionItem__description" ng-if="item.type !== 'recentSearch'" ng-bind-html="item.description"></div>
```
Why do we need to use `ng-bind-html` on either of the two lines above? Wouldn't `{{}}` work? `ng-bind-html` will sanitize the value so fields like `xss` in makelogs will show an empty string as a suggestion. `{{}}` will escape the html characters so the field value still displays correctly.

Not 100% sure but it also seemed like switching to `{{}}` fixed the console errors I mentioned in my summary comment.

Bargs replied:
I see now that the suggestion descriptions can have html in them. But won't this cause issues if the field names and/or values contain html characters? It seems like we either need to html escape all user provided values in the suggestion modules themselves before returning the suggestion objects, or refactor the suggestion objects so they don't need to contain html.

Bargs commented on src/ui/public/query_bar/directive/query_bar.js:
```diff
@@ -36,18 +45,47 @@ module.directive('queryBar', function () {
+        this.updateSuggestions();
+      };
+
+      $scope.$watch('queryBar.localQuery.language', (language) => {
```
I think there's a copy paste error here? Line 95 has an almost identical watcher.

Bargs commented on src/ui/public/query_bar/directive/query_bar.js:
```diff
@@ -36,18 +45,47 @@ module.directive('queryBar', function () {
-        this.typeaheadItems = persistedLog.get().filter(recentSearch => {
-          return recentSearch.includes(query) && recentSearch !== query;
+      this.updateSuggestions = () => {
+        $timeout(() => {
```
Why does this need to run in a `$timeout`?

w33ble replied:
I was curious about that too. I figured it was because the underlying service wasn't using the angular promise library, but it's also calling `$scope.$apply`, so I wouldn't think this needed to be in a `$timeout` as well.

Bargs commented on src/core_plugins/kibana/public/discover/index.html:
```diff
@@ -28,6 +28,7 @@ <h1 tabindex="0" id="kui_local_breadcrumb" class="kuiLocalBreadcrumb">
           query="state.query"
           app-name="'discover'"
           on-submit="updateQueryAndFetch($query)"
+          index-pattern="indexPattern"
```
We need to make the same change on visualize and dashboard. What should happen on dashboards where there could be 0 or more index patterns? 

We also probably need to handle the case where no index pattern is passed more gracefully, if we want to support that use case, because right now it spams the console with errors.

Bargs commented on src/ui/public/query_bar/lib/match_pairs.js:
```diff
@@ -0,0 +1,49 @@
+ *    and the characters before and after the cursor correspond to a pair,
+ *    remove both characters and move the cursor backward.
+ */
+export const matchPairsProvider = ({
```
I don't see this file actually being used anywhere.

Bargs commented on src/ui/public/directives/match_pairs.js:
```diff
@@ -0,0 +1,101 @@
+
+      function moveSelectionForward() {
+        e.preventDefault();
+        target.setSelectionRange(selectionStart + 1, selectionEnd + 1);
```
Is this working correctly? Every time I create a selection that ends at a closer, then type the closer, it just deletes the selection (in other words, the default browser behavior). 

![moveselectionforwardissue](https://user-images.githubusercontent.com/6239176/37351704-07af8772-26b2-11e8-85c7-b88cb16638bc.gif)

Bargs commented on src/ui/public/directives/match_pairs.js:
```diff
@@ -0,0 +1,101 @@
+        if (precedingCharacter === '\\') return false;
+
+        // Don't insert if it's a quote and the either of the preceding/following characters is alphanumeric
+        return !(['"', `'`].includes(key) && (isAlphanumeric(precedingCharacter) || isAlphanumeric(followingCharacter)));
```
Can you explain why this is necessary? It's not apparent to me.

Bargs commented on src/ui/public/directives/match_pairs.js:
```diff
@@ -0,0 +1,101 @@
+        const followingCharacter = value.charAt(selectionStart + 1);
+
+        // Don't insert if the preceding character is a backslash
+        if (precedingCharacter === '\\') return false;
```
This makes me think this module is somewhat coupled to our KQL syntax. Should we rename and move the file to make that more apparent since it's not totally generic?

Bargs commented on src/ui/public/kuery/suggestions/conjunction.js:
```diff
@@ -0,0 +1,22 @@
+const type = 'conjunction';
+
+const conjunctions = {
+  and: `<p>Requires that <span class="suggestionItem__callout">both arguments</span> joined together both return true</p>`,
```
This reads a little strangely to me what about just `Requires both arguments to be true`?

Bargs commented on src/ui/public/kuery/suggestions/conjunction.js:
```diff
@@ -0,0 +1,22 @@
+
+const conjunctions = {
+  and: `<p>Requires that <span class="suggestionItem__callout">both arguments</span> joined together both return true</p>`,
+  or: `<p>Requires <span class="suggestionItem__callout">one or more of</span> the arguments joined together must return true</p>`
```
`Requires one or more of the arguments to be true`?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/587/)

lukasolson commented:
@AlonaNadler 

> can  limit the number of suggestions

Yeah, we can limit the number of suggestions to any arbitrary limit we decide... But I'm not sure how we decide what that number is... Varying screen size may be able to display variable number of suggestions.

> Any chance the suggestions won't be case sensitive e.g. when typing `Win 7` instead of `win 7` might cause confusion `geo.dest : CN` vs `geo.dest : cn` - I kinda know what you going to tell me ... :)

Yeah, I think you know what I'm going to say ;) The way the Elasticsearch query is set up prevents us from doing case insensitive suggestions, or at least doing so efficiently.

> Maybe worth considering hovering on the first suggestion as default, currently to select the suggestions I need to use the arrows and return, if I want to select the first suggestion intuitively I click return (to select it ) but then it triggers the search

We can discuss this, but I think if you had to hit escape every time before you execute a search it would be less user friendly than having to select suggestions.

w33ble commented on src/ui/public/kuery/suggestions/value.js:
```diff
@@ -0,0 +1,48 @@
+  `;
+}
+
+export function getSuggestionsProvider({ $http, config, indexPattern }) {
```
This seems to be the only suggestion provider using `$http` and `config`. 

I don't think you need to use `$http` at all, since you're [wrapping the response in a native promise upstream](https://github.com/lukasolson/kibana/blob/f2a0832e13ceaf5c2292e96abe3de2961bb4fc21/src/ui/public/kuery/suggestions/index.js#L27) anyway, and then [later using $scope.$apply](https://github.com/elastic/kibana/pull/15945/files/f2a0832e13ceaf5c2292e96abe3de2961bb4fc21#diff-a63e1054ef0dce774fee5c3a7c450184R64). `Fetch` or some wrapper would give you the same functionality, and remove the Angular dependency here.

And the only thing you're using the `config` service for is to check the value of `filterEditor:suggestValues`. Seems like it would be better to get this (and any other config value you might need) upstream and pass down a config object of some kind. It'll make this code easier to decouple from Angular in the future too, no? This one may be a bit moot since you still need to `indexPattern` and `persistedLog` services anyway...

w33ble commented on src/ui/public/query_bar/directive/query_bar.js:
```diff
@@ -15,39 +19,74 @@ module.directive('queryBar', function () {
     bindToController: true,
-    controller: callAfterBindingsWorkaround(function ($scope, config, PersistedLog) {
+
+    controller: callAfterBindingsWorkaround(function ($scope, $element, $http, $timeout, config, PersistedLog) {
```
You're only pulling `$http` in here so you can pass it into the suggestions provider. As noted elsewhere, you don't need it at all, but you could also better encapsulate it by creating a simple kuerySuggestor factory (or service). You could also avoid passing in the config service externally that way as well.

`/src/ui/public/query_bar/service/kuery_suggest.js`
```js
import { uiModules } from 'ui/modules';
import { getSuggestionsProvider } from '../../kuery';

const module = uiModules.get('kibana');

module.factory('kuerySuggestor', function ($http, config) {
  return ({ persistedLog, indexPattern }) => getSuggestionsProvider({ $http, config, persistedLog, indexPattern });
});
```

Then here, you can just add `kuerySuggestor`, and down where you use it, you'd just have `getSuggestions = kuerySuggestor({ indexPattern, persistedLog });`

You could almost certainly take this even further and encapsulate the persisted log part there with a little refactoring of the factory, instead of mixing that concern into the `queryBar` directive here. ie) `getSuggestions.addHistory(this.localQuery.query);`

Further still, you could probably remove the upstream `$scope.$apply` stuff by wrapping `getSuggestionsProvider` in the Angular `Promise` service...

```js
import { uiModules } from 'ui/modules';
import { getSuggestionsProvider } from '../../kuery';

const module = uiModules.get('kibana');

module.factory('KuerySuggestor', function ($http, config, PersistedLog, Promise) {
  return class KuerySuggestor {
    constructor(persistedName, indexPattern) {
      this.persistedLog = new PersistedLog(persistedName, {
        maxLength: config.get('history:limit'),
        filterDuplicates: true
      });
      this.indexPattern = indexPattern;
    }

    getSuggestions() {
      return Promise.resolve(
        getSuggestionsProvider({
          $http,
          config,
          persistedLog: this.persistedLog,
          indexPattern: this.indexPattern
        })
      );
    }

    addHistory(entry) {
      this.persistedLog.add(entry);
    }
  };
});

// controller code:
controller: callAfterBindingsWorkaround(function ($scope, $element, $timeout, config, KuerySuggestor) {

// later in the watcher:
kuerySuggestor = new KuerySuggestor(`typeahead:${this.appName}-${language}`, indexPattern);

// and persisting the query...
this.submit = () => {
  if (this.localQuery.query) kuerySuggestor.add(this.localQuery.query);
  this.onSubmit({ $query: this.localQuery });
};
```

Now the Angularized concerns all live in one place, and the dependencies do too.

w33ble replied:
This would also give you a fine place to decouple the config service and just pass in a config object...

```js
this.suggestionProvider = getSuggestionsProvider({
  $http,
  config: {
    suggestValues: !!config.get('filterEditor:suggestValues'),
  },
  persistedLog: this.persistedLog,
  indexPattern: this.indexPattern
});
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/605/)

w33ble commented on src/ui/public/directives/match_pairs.js:
```diff
@@ -0,0 +1,101 @@
+      const { target, key, metaKey } = e;
+      const { value, selectionStart, selectionEnd } = target;
+
+      if (shouldMoveCursorForward()) {
```
I hate to be that guy, but you should probably [return early](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#returnthrow-early-from-functions).

```js
if (shouldMoveCursorForward()) return moveCursorForward();
if (shouldInsertMatchingCloser()) return insertMatchingCloser();
if (shouldRemovePair()) return removePair();
```

w33ble commented on src/ui/public/directives/match_pairs.js:
```diff
@@ -0,0 +1,101 @@
+        removePair();
+      }
+
+      function shouldMoveCursorForward() {
```
All these functions should be hoisted out of the event handler, and you should just pass the key in. Defining them (and garbage collecting them) every time there's a key press is excessive.

lukasolson replied:
This is pretty simple for the `should...` functions but it seems odd for the others since they need `ngModel`. I pushed a commit just now... Let me know what you think: 57159424e2

w33ble replied:
Yeah, that commit looks good. 👍

w33ble commented on packages/kbn-ui-framework/dist/ui_framework.css:
```diff
@@ -2730,6 +2730,8 @@ main {
     .theme-dark .kuiLocalSearchInput:focus {
       outline: none;
       border-color: #0079a5; }
+  .kuiLocalSearchInput:focus {
```
I assume this is why the focus indication on the input look different than all the other elements on the page. Should it be styled to look the same?

![mar-13-2018 17-19-31](https://user-images.githubusercontent.com/404731/37376867-bc03a254-26e3-11e8-924c-aa8057462bef.gif)

lukasolson replied:
@snide?

snide replied:
It had a double outline on this one specifically which was pretty gnarly. I spot fix really bad cases, but ultimately a lot of these issues need to be fixed through EUIfication. Till then I don't plan on updating the accessibility focus styles (which are a pretty aggressive css selector) since we want all this stuff to go out the door.

w33ble replied:
Is it worth moving this to EUI then? Even if we don't use the React components here, perhaps we can use the classes in Angular? Perhaps that's pretty brittle though, I don't know how stable the underlying CSS is...

FWIW, I'm not proposing blocking this PR on it, it's OK, just a little unfortunate that the appearance isn't consistent.

snide replied:
The whole thing is pretty brittle and brute forced, for example... here's what the `.less` side of things looks like for kibana when it comes to focus states.

```
:focus {
  &:not([class^="eui"]),
  {
    @include focus;
  }
}
```

Basically, all elements that need some sort of focus that aren't eui based (which has focus / accessibility as a core consideration) get a brute force attack that does the weird double outline (because who knows what the background color of the element is!). It's the only way that I knew to make 6.0 accessible quickly. Obviously this is a terrible way to do things and it makes the visuals totally inconsistent, because if you _don't_ want that brute force style, or if they already had some sort of focus state, they'll be off. So in practice, I've avoided doing that (except in rare cases), knowing that this stuff will automatically get fixed with EUI adoption.

TL;DR

We can do it, but it's brittle and we should just switch stuff over (not in this PR).

w33ble commented on src/ui/public/kuery/suggestions/conjunction.js:
```diff
@@ -0,0 +1,22 @@
+      const description = getDescription(conjunction);
+      return { type, text, description, start: end, end };
+    });
+    return Promise.resolve(suggestions);
```
Wrapping all the suggestion results in a promise isn't necessary. They all get consumed through `getSuggestionsProvider` in `src/ui/public/kuery/suggestions/index.js` anyway, which wraps the results in a promise anyway, so returning a promise of a value doesn't change how it works when it's actually being consumed. These are all "private" by convention, so they aren't used directly.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/641/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/668/)

lukasolson commented:
@w33ble:

> I find it odd that clicking on a previous query doesn't do anything more than update the input

Fixed in 37b495687d8165f8ac0c81d8552dcc67373f785a.

> Also related, it's odd to me that when the form is submitted that the auto-complete dialog doesn't go away.

Fixed in 3271f54adac48efd01a96a0edf90d425a4bcdbf0.

> Keyboard navigation will need to cause the option container to scroll up and down.

Fixed in 7f59d35173890ab3a1256d17aab279b942e9ca33.

> The hide/show state seems to be preserved on the input. Initially it's hidden, until the user types something, which seems correct. When the input loses focus, however, the show state is restored immediately when it gains focus again.

Fixed in 790f1791c9.

w33ble commented:
@lukasolson latest commits are a huge improvement! Interestingly, it only solves a couple of the things I pointed out, but it's a whole lot nicer now. Here's a list of what I reported and where they stand, for reference:

- [x] Clicking previous query does not execute it
- [ ] Clicking/enter on previous query does not close auto-complete dialog
- [ ] Enter to execute does not close auto-complete dialog
- [x] Keyboard nav does not scroll the auto-complete dialog
- [ ] Hide/show state is preserved even when input loses focus

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/675/)

w33ble commented:
One last weird thing. If you use parens in the value, you lose the value recommendation until you type a sting that matches:

![mar-15-2018 14-59-00](https://user-images.githubusercontent.com/404731/37493515-66e968aa-2862-11e8-9509-80d2a49050b6.gif)

Here, it recommends the values from `extension.raw`, but once I type a paren, that goes away. When I type a quote, I still get nothing, and it's only once I type a character that matches a value that I get some suggestions.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/683/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/822/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/832/)

Bargs commented on src/ui/public/query_bar/directive/query_bar.js:
```diff
@@ -15,55 +20,107 @@ module.directive('queryBar', function () {
+
+      this.getIndexPatterns = () => {
+        if (compact(this.indexPatterns).length) return Promise.resolve(this.indexPatterns);
+        return Promise.all([indexPatterns.getDefault()]);
```
Is it necessary to grab the default here? Won't each individual app do that if it's the right thing to do? I was thinking if no indexPattern was passed in, we just wouldn't provide suggestions. If we didn't have to grab the default, I think it would remove the need to deal with promises here?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/843/)

w33ble commented on src/ui/public/kuery/suggestions/value.js:
```diff
@@ -0,0 +1,48 @@
+export function getSuggestionsProvider({ $http, config, indexPatterns }) {
+  const allFields = flatten(indexPatterns.map(indexPattern => indexPattern.fields.raw));
+  const shouldSuggestValues = config.get('filterEditor:suggestValues');
+  return shouldSuggestValues ? memoize(getValueSuggestions, getFieldQueryHash) : () => Promise.resolve([]);
```
I like this, but I think adding some kind of TTL here would be a good idea. The cache only gets cleared when users refresh the browser or navigate to an app outside of core, which could be never, so new indexed values won't show up. 

A TTL of, say, a minute, would calm the lookup requests to ES while also allowing new values to show up over time.

w33ble commented:
> After clicking on a suggestion, I don’t get any more suggestions until I type something

@Bargs which suggestion? This seems like the correct behavior for historical queries, but the incorrect one for all the other suggestions...

Bargs commented:
@w33ble all suggestions

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/886/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/898/)

w33ble commented:
@lukasolson Latest dialog open/close handling looks golden! 🏅

w33ble commented on src/ui/public/query_bar/directive/query_bar.js:
```diff
@@ -15,55 +20,106 @@ module.directive('queryBar', function () {
-          return recentSearch.includes(query) && recentSearch !== query;
-        });
+      this.updateSuggestions = () => {
+        const { query } = this.localQuery;
```
It's possible for `query` to be undefined here, and get passed into the suggestion code, which will throw. This needs a check. You can recreate the issue by removing the `query.query` part section in the URL, which modifies appState, and thus `this.localQuery`.

w33ble replied:
It'll also throw another error if you set `query:()` in the appState. The resulting `_msearch` fails. Here's an example query in case it's helpful.

```
{"index":["logstash-*"],"ignore_unavailable":true,"preference":1521590260741}
{"version":true,"size":500,"sort":[{"@timestamp":{"order":"desc","unmapped_type":"boolean"}}],"_source":{"excludes":[]},"aggs":{"2":{"date_histogram":{"field":"@timestamp","interval":"30m","time_zone":"America/Phoenix","min_doc_count":1}}},"stored_fields":["*"],"script_fields":{},"docvalue_fields":["@timestamp","relatedContent.article:modified_time","relatedContent.article:published_time","utc_time"],"query":{"bool":{"must":[{},{"range":{"@timestamp":{"gte":1520751600000,"lte":1520818649832,"format":"epoch_millis"}}}],"filter":[],"should":[],"must_not":[]}},"highlight":{"pre_tags":["@kibana-highlighted-field@"],"post_tags":["@/kibana-highlighted-field@"],"fields":{"*":{}},"fragment_size":2147483647}}
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/934/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/936/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/938/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/942/)

Bargs commented on src/ui/public/kuery/suggestions/value.js:
```diff
@@ -0,0 +1,49 @@
+}
+
+function resolver(query, field) {
+  return `${query}/${field.name}`;
```
Should this include the index pattern id as well? Otherwise I think two different index patterns with a field that happens to have a matching name could get stale suggestions.

lukasolson replied:
Totally should

Bargs commented:
Pending a green CI, this looks great to me!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/957/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/973/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/977/)

lukasolson commented:
6.x (6.3.0): b665a30

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1004/)
